### ISSUE TITLE
One-to-One Association keeps old records

### ISSUE THREAD

**Issue by mereghost** [rom-rb/rom-sql#205]
Heya folks,

I was working on `hanami/model` support for [one-to-one associations](https://github.com/hanami/model/blob/has_one/lib/hanami/model/associations/has_one.rb) when I noticed that when I associate a new record the old one is kept in the database.

What I expected was that issuing a create command on a one-to-one association would remove the old record (if any) and insert a new one. While it does retrieve the correct one, it leaves the database a mess and might generate constraint violation errors.

**Comment by solnic** [comment]
Thanks for reporting this. We'll address this in the next version.

**Comment by mereghost** [comment]
By next version you mean rom-4.0? =)

**Comment by solnic** [comment]
Nope, next bug fix release. @flash-gordon you think you could take care of this?

**Comment by flash-gordon** [comment]
@solnic sorry for being stupid here, but how is it supposed to work? Remove existing references [in a before hook](https://github.com/rom-rb/rom-sql/blob/2d09761ee07e831dc65ad1a9f17531ebfa9cc2df/lib/rom/sql/plugin/associates.rb#L70-L89)?
@mereghost I'd say it should trigger a constraint violation error, even if we remove the existing record this operation is not going to be atomic, leaving room for possible duplications. FWIW I'd recommend creating one-to-one records along simultaneously, in a single transaction, and use updates later on.

**Comment by solnic** [comment]
@flash-gordon I thought about wrapping it in a transaction as a special way of handling this type of association

**Comment by flash-gordon** [comment]
yes, this should be run inside a transaction, so that possible errors won't pollute the database with partial writes, but this won't prevent from having duplicates anyway, consider this

session 1 | session 2
---------|---------
starts a command | starts a command
deletes associated records -> 0 | ...
... | deleted associated records -> 0
inserts a new record | ...
... | inserts a new record
commit | commit

Having a unique constraint will pause session 2 on the insert step until session 1 is completed (or rolled back). So we actually need a DB constraint. And yes, transaction is a must too :)
Another option is using pessimistic locking on the parent record, i.e. `SELECT FOR UPDATE`, this way both sessions will be synchronized before the delete phase starts. Unfortunately, not all databases support this mechanism, for instance, SQL does not.

**Comment by mereghost** [comment]
@flash-gordon While I do agree that having a constraint at the database level is the right thing to do, I see no way to enforce it to the end users.

Guarding against partial writes and then warning users that duplicates *may* end in the database, due to race conditions, if no unique constraint is in place seems more "user friendly" to me.
